### PR TITLE
[GH-22] Condense all array types into single Array class

### PR DESCRIPTION
**Pull request by fwkoch**
This array class stores size and standard array type. Boolean arrays are also now allowed.

### CODE REVIEW

File: omf/data.py [hunk 1]
```diff
@@ -130,89 +165,16 @@ def _check_limits_on_validate(self):
...
-        'text at locations on a mesh (see location parameter)',
-        StringArray,
-    )
-
-
-class DateTimeData(ProjectElementData):
-    """Data array with DateTime entries"""
-    array = properties.Instance(
-        'datetimes at locations on a mesh (see location parameter)',
-        DateTimeArray,
```
Comment (dbrookes96) [inline omf/data.py:211]: What is the plan for string, datetime and color data types? I guess the intention is to use the new types to represent these. i.e. color would be a 3 x N array of uint8. What about strings?
I think I may have mentioned this in the past how things like strings, datetime and color shouldn't be represented in the JSON to try and keep the project.json as tiny and fast to parse as possible. Is that the case?

File: omf/data.py [hunk 2]
```diff
@@ -11,16 +11,36 @@
...
+    'Uint8Array': np.dtype('uint8'),
+    'Int16Array': np.dtype('int16'),
+    'Uint16Array': np.dtype('uint16'),
+    'Int32Array': np.dtype('int32'),
+    'Uint32Array': np.dtype('uint32'),
+    'Int64Array': np.dtype('int64'),
+    'Uint64Array': np.dtype('uint64'),
+    'Float32Array': np.dtype('float32'),
+    'Float64Array': np.dtype('float64'),
+    'BooleanArray': np.dtype('bool'),
```
Comment (dbrookes96) [inline omf/data.py:27]: No string type?

File: omf/data.py [hunk 3]
```diff
@@ -222,19 +184,21 @@ class Legend(ContentModel):
...
         props=(
-            ColorArray,
-            DateTimeArray,
-            StringArray,
-            ScalarArray,
+            properties.List('', properties.Color('')),
+            properties.List('', properties.String('')),
+            properties.List('', properties.Integer('')),
+            properties.List('', properties.Float('')),
         )
```
Comment (dbrookes96) [inline omf/data.py:265]: Are color and string type limited only to MappedData Legends?

File: omf/serializers.py [hunk 4]
```diff
@@ -41,37 +41,17 @@ def array_serializer(arr, open_file, **kwargs):                                #
...
-            assert abs(unknown_dim - int(unknown_dim)) < 1e-9, 'bad shape'
-            shape = tuple(
-                (int(unknown_dim) if dim == '*' else dim for dim in self.shape)
-            )
-        else:
-            assert abs(unknown_dim - 1) < 1e-9, 'bad shape'
-            shape = self.shape
-        arr = arr.reshape(shape)
-        return arr
+    assert index['dtype'] in ('<i8', '<f8'), 'invalid dtype'
```
Comment (dbrookes96) [inline omf/serializers.py:49]: ~~How are these array data types going to be represented in the JSON? Will we still use the '<i8', '<f8', etc. numpy strings or will we just rely on the __class__ names like "Int64Array", "Float64Array".~~

Comment (dbrookes96) [reply omf/serializers.py:49]: I see the test output uses "Array" as the type so I assume the '<i8' esque strings will be used.

File: omf/data.py (hunk 1 above)
Comment (fwkoch) [reply omf/data.py:211]: Picking up after almost 2 years... hah. Made a few changes to address your concerns:

- Color data: You may use MappedData + Legend or N x 3 as uint8 as you suggest. No explicit color data.
- String / Datetime data: There is a separate array class for these. Both are just lists of strings, but the `datatype` will specify if every entry is a valid datetime. My thinking is these will be serialized to a separate file (so they don't blow up the base JSON), but probably just as another JSON blob.

File: omf/serializers.py (hunk 4 above)
Comment (fwkoch) [reply omf/serializers.py:49]: This will be resolved with file serialization... https://github.com/gmggroup/omf/pull/91

**Pull request merged**